Configurable openshift component logging

charlesakalugwu opened the pull request:
This work makes the log level of the openshift components configurable for clusters.

This currently only allows changing the log levels for api-server, node and controller-manager components to start with. 

This is part of AZURE-257

```release-note
plugin: geneva action to change openshift component level logging
```

/cc @jim-minter @kwoodson @thekad @asalkeld

kwoodson commented on pkg/api/config.go:
```diff
@@ -11,6 +11,8 @@ import (
 	// ClusterVersion defines release version of the plugin used to build the cluster
 	ClusterVersion string `json:"clusterVersion,omitempty"`
+	// ClusterLogLevel specifies the OpenShift component log level
+	ClusterLogLevel uint8 `json:"clusterLogLevel,omitempty"`
```
Would it be better to make this a const?  Similar to https://github.com/openshift/openshift-azure/blob/master/pkg/api/types.go#L71

charlesakalugwu replied:
good idea

kwoodson commented on pkg/api/validate/plugin.go:
```diff
@@ -24,6 +24,10 @@ func (v *PluginAPIValidator) Validate(t *pluginapi.Config) (errs []error) {
 		errs = append(errs, fmt.Errorf("invalid ClusterVersion %q", t.ClusterVersion))
 	}
+	// validate ClusterLogLevel
+	if t.ClusterLogLevel%2 != 0 || (t.ClusterLogLevel < 2 || t.ClusterLogLevel > 8) {
```
0 is a valid clusterloglevel.  Any reason we are excluding it?

charlesakalugwu replied:
rationale was that setting 0 wouldn't be helpful to a cluster admin. I'll revert this

kwoodson commented on pkg/arm/data/node-startup.sh:
```diff
@@ -9,7 +9,7 @@ if [ -f "/etc/sysconfig/atomic-openshift-node" ]; then
 
 # remove registry certificate softlink from docker
-unlink /etc/docker/certs.d/registry.access.redhat.com/redhat-ca.crt 
+unlink /etc/docker/certs.d/registry.access.redhat.com/redhat-ca.crt
```
not for this PR but this should move to the node image build.

jim-minter commented on pkg/api/config.go:
```diff
@@ -11,6 +11,8 @@ import (
 	// ClusterVersion defines release version of the plugin used to build the cluster
 	ClusterVersion string `json:"clusterVersion,omitempty"`
+	// ClusterLogLevel specifies the OpenShift component log level
+	ClusterLogLevel uint8 `json:"clusterLogLevel,omitempty"`
```
This is going back on what I said to you before, but it's little extra work: LogLevel would be better as a struct containing different components and their log levels.  Example: apiserver, controllers, node.  Can use an int rather than a uint8.

asalkeld replied:
yes, otherwise you get all or nothing.

charlesakalugwu replied:
good idea. I'll get this in

jim-minter commented on pkg/fakerp/admin_handlers.go:
```diff
@@ -124,3 +126,45 @@ func (s *Server) handleRotateSecrets(w http.ResponseWriter, req *http.Request) {
 }
+
+// handleGetLogLevel handles admin requests to get the configured OpenShift component log level for OpenShiftManagedClusters
+func (s *Server) handleGetLogLevel(w http.ResponseWriter, req *http.Request) {
```
you don't need any fakerp changes or additional geneva action - with the config struct in place the generic admin API can handle getting and putting loglevel changes.

charlesakalugwu replied:
alright

jim-minter commented on pluginconfig/pluginconfig-311.yaml:
```diff
@@ -13,6 +13,14 @@
+  # 4 - Debugging-level information
+  # 6 - API-level debugging information (request / response)
+  # 8 - Body-level API debugging information
+clusterLogLevel: 4
```
default to 2, not 4

charlesakalugwu replied:
done

jim-minter commented on test/util/populate/walk.go:
```diff
@@ -115,7 +115,7 @@ func (w walker) walk(v reflect.Value, path string) {
 		v.SetInt(1)
 	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
-		v.SetUint(1)
+		v.SetUint(2)
```
revert

charlesakalugwu replied:
done

thekad commented on pkg/api/validate/plugin_test.go:
```diff
@@ -14,6 +14,7 @@ func TestPluginTemplateValidate(t *testing.T) {
 			errors.New(`imagePublisher should be redhat`),
 			errors.New(`invalid ImageSKU ""`),
 			errors.New(`invalid ClusterVersion ""`),
+			errors.New(`invalid ClusterLogLevel 0`),
```
log level 0 is valid afaik

charlesakalugwu replied:
yep

thekad commented on pkg/arm/data/master-startup.sh:
```diff
@@ -44,9 +44,7 @@ fi
-{{- if $.Extra.TestConfig.RunningUnderTest }}
-sed -i -e "s#DEBUG_LOGLEVEL=2#DEBUG_LOGLEVEL=4#" /etc/sysconfig/${SERVICE_TYPE}-node
-{{- end }}
+sed -i -e "s#DEBUG_LOGLEVEL=2#DEBUG_LOGLEVEL={{ .Config.ClusterLogLevel }}#" /etc/sysconfig/${SERVICE_TYPE}-node
```
I think with these changes we're losing the ability to get debug logs when `RUNNING_UNDER_TEST`, I'd suggest work that back into the dev process, either by defaulting the logLevel to 4 when running under test, or exposing the logLevel to be configurable through environment variables.

charlesakalugwu replied:
i can readd this in a follow-up

charlesakalugwu replied:
@jim-minter just to make sure, are you ok with @thekad 's suggestion? I think I prefer his second option. what do you think?

charlesakalugwu commented:
/retest

jim-minter commented on pkg/api/convertertoadmin.go:
```diff
@@ -104,6 +104,7 @@ func ConvertToAdmin(cs *OpenShiftManagedCluster) *admin.OpenShiftManagedCluster
 func convertConfigToAdmin(cs *Config) *admin.Config {
 	return &admin.Config{
 		ClusterVersion:                       &cs.ClusterVersion,
+		ClusterLogLevel:                      convertClusterLogLevelsToAdmin(cs.ClusterLogLevel),
```
there are no changes made to convertfromadmin.go and there should be.  This should have triggered a unit test failure.  Add a unit test that does a round trip: populate a dummy `*admin.OpenShiftManagedCluster`, convert it to a `*OpenShiftManagedCluster`, convert that back again and check that what you get equals what you started with.

charlesakalugwu replied:
thats right, theres even more stuff that had been left out in conversions

```
=== RUN   TestRoundTrip
--- FAIL: TestRoundTrip (0.00s)
    converterfromadmin_test.go:301: unexpected diff [Config.ClusterVersion: Config.ClusterVersion !=  Config.ClusterLogLevel.ApiServer: 1 != 0 Config.ClusterLogLevel.ControllerManager: 1 != 0 Config.ClusterLogLevel.Node: 1 != 0 Config.GenevaLoggingControlPlaneEnvironment: Config.GenevaLoggingControlPlaneEnvironment !=  Config.GenevaLoggingControlPlaneRegion: Config.GenevaLoggingControlPlaneRegion != ]
```

I'll add them in

jim-minter commented on pkg/api/validate/plugin.go:
```diff
@@ -53,6 +55,19 @@ func (v *PluginAPIValidator) Validate(t *pluginapi.Config) (errs []error) {
 }
 
+func (v *PluginAPIValidator) validatePluginTemplateComponentLogLevels(c *pluginapi.ComponentsLogLevel) (errs []error) {
+	if c.ApiServer%2 != 0 || (c.ApiServer < 0 || c.ApiServer > 8) {
```
the %2 does not make sense.  Make the types uints, then check they are not larger than, say, 20.

charlesakalugwu replied:
valid logging levels are 0,2,4,6 and 8. The %2 checks that the supplied log level is an even number. Am i missing something?

jim-minter commented on pluginconfig/pluginconfig-311.yaml:
```diff
@@ -13,6 +13,17 @@
 ## Geneva integration other details
 clusterVersion: v0.0
+## Openshift component logging levels
+#levels:
```
this is not the right place for this documentation - please remove.

charlesakalugwu replied:
alright

jim-minter commented on test/e2e/specs/fakerp/loglevel.go:
```diff
@@ -0,0 +1,58 @@
+		Expect(after).NotTo(BeNil())
+
+		By("Verifying that the cluster log level has been updated")
+		Expect(after.Config.ClusterLogLevel).To(Equal(before.Config.ClusterLogLevel))
```
I don't understand why this is succeeding.  I would expect it currently to fail because of validateUpdateConfig (which this PR should be updating).  That's a fairly serious bug - please solve it before this gets merged.

charlesakalugwu replied:
Having a look. I also have a PR for the prow jobs open: https://github.com/openshift/release/pull/2914 . Can we get that merged so I can run `/test changeloglevel` in CI?

charlesakalugwu commented:
/hold

charlesakalugwu commented:
/test unit

charlesakalugwu commented:
/retest unit

charlesakalugwu commented:
/retest

jim-minter commented on pkg/api/admin/api/config.go:
```diff
@@ -6,9 +6,18 @@ import (
 	// ClusterVersion (out) defines release version of the plugin used to build the cluster
 	ClusterVersion *string `json:"clusterVersion,omitempty"`
+	// ClusterLogLevel specifies the log levels for the various openshift components
+	ClusterLogLevel *ComponentsLogLevel `json:"clusterLogLevel,omitempty"`
```
yikes - pick one name and stick to it
```
Foo *Foo `json:"foo,omitempty"`
```

jim-minter replied:
also - use spaces to separate logical groups

jim-minter commented on pkg/api/converterfromadmin.go:
```diff
@@ -349,9 +354,27 @@ func mergeConfig(oc *admin.OpenShiftManagedCluster, cs *OpenShiftManagedCluster)
 	if in.GenevaLoggingControlPlaneAccount != nil {
 		out.GenevaLoggingControlPlaneAccount = *in.GenevaLoggingControlPlaneAccount
 	}
+	if in.GenevaLoggingControlPlaneEnvironment != nil {
```
your new tests found these were missing?  If so I owe you :beer:

charlesakalugwu replied:
@jim-minter yes the test discovered the following were missing:

```
Config.GenevaLoggingControlPlaneEnvironment
Config.GenevaLoggingControlPlaneRegion
```

jim-minter commented:
@charlesakalugwu also rebased and pushed - ptal
please can you check the test is working - lmk then will lgtm
/hold

charlesakalugwu commented:
/test changeloglevel

charlesakalugwu commented:
/retest

jim-minter commented:
retest broke because https://github.com/openshift/openshift-azure/pull/1220 is in flight, will try retest again once that lands

jim-minter commented:
/retest

jim-minter commented:
/test changeloglevel

mjudeikis commented on pkg/api/config.go:
```diff
@@ -68,6 +72,13 @@ type Config struct {
 }
 
+// ComponentLogLevel represents the log levels for the various components of a cluster
+type ComponentLogLevel struct {
```
Router is needed too.

mjudeikis replied:
Can be follow-up

mjudeikis commented on pkg/api/converterfromadmin.go:
```diff
@@ -300,7 +300,12 @@ func convertIdentityProviderAdmin(in admin.IdentityProvider, old *IdentityProvid
 func mergeConfig(oc *admin.OpenShiftManagedCluster, cs *OpenShiftManagedCluster) {
 	in, out := oc.Config, &cs.Config
-
+	if in.ClusterVersion != nil {
```
Why this one was added? I think we dont propagate this field back to OpenShiftManageCluster?

mjudeikis replied:
Or my brains still on holiday?

jim-minter replied:
this PR added a round trip test to prove that you can convert admin -> internal -> admin and get the same thing back - this discovered several missing fields which are added here

mjudeikis commented:
@openshift/sig-azure let's stop merging release jobs before PR is merged/or we are 100% sure tests are passing. 
We can test these locally using:https://github.com/openshift/release/tree/master/projects/azure#testing-prow-jobs

charlesakalugwu commented:
@jim-minter I discovered and fixed the following issues and fixed those back up to your commit.

- Json tags for componentLogLevel was incorrectly specified as: 

```
ComponentLogLevel ComponentLogLevel `json:"componentLogLevels,omitempty"`
```
- I added a TestJsonTags test to check for the conformance of all the plugin template's json tags
- Changed e2e tests to perform relative updates of the log level. Instead of setting it to 4 (which results in a no-op update on clusters created with RUNNING_UNDER_TEST=true) we subtract 2 from the current log levels. As we run e2e tests on fresh clusters this should not be a problem. 
- fakerp.overridePluginTemplate now also accepts a boolean isUpdate argument. This allows us to only set loglevel=4 to clusters being created under test. We should be able to update loglevel afterwards.
- Changed the following sed commands in the node startup scripts from 
```
sed -i -e "s#DEBUG_LOGLEVEL=2#DEBUG_LOGLEVEL={{ .Config.ComponentLogLevel.Node }}#" /etc/sysconfig/${SERVICE_TYPE}-node
```
to
```
sed -i -e "s#DEBUG_LOGLEVEL=.*#DEBUG_LOGLEVEL={{ .Config.ComponentLogLevel.Node }}#" /etc/sysconfig/${SERVICE_TYPE}-node
```
This is to ensure that the change takes effect always instead of only on clusters with loglevel=2 (clusters created with RUNNING_UNDER_TEST=false)
- updated validateUpdateContainerService to allow updates to the log level.

charlesakalugwu commented:
/test changeloglevel

jim-minter commented:
thanks @charlesakalugwu 
/hold cancel
/lgtm

charlesakalugwu commented:
/retest

charlesakalugwu commented:
/retest

jim-minter commented:
/lgtm cancel
e2e issue looks genuine

charlesakalugwu commented:
/test changeloglevel

asalkeld commented:
needs rebase

jim-minter commented:
closing for #1271
